In [ ]:
!dir "../pantho_Tufano_processing/processed/acumos_12"

In [ ]:
BASE="../pantho_Tufano_processing/processed/acumos_12"
main_BASE = "../codes/acumos_12"
code = open(BASE+"/1.java", 'r', encoding = 'utf-8')
main_code = open(main_BASE+'/1.java', 'r', encoding = 'utf-8').read()
mp = open(BASE+"/1.java.map", 'r', encoding = 'utf-8')
code = code.read()
mp = mp.read()
c = 0
keys = []
swap1 = []
swap2 = []
dic = {}
for val in mp.split('\n'):
    if len(val)>0:
        if c%2==0:
            swap1 = val.split(',')
        else:
            swap2 = val.split(',')
            if len(swap1) != len(swap2):
                print("map length mismatch ")
            for i in range(len(swap2)):
                #code = code.replace(swap2[i], swap1[i])
                dic[swap2[i]] = swap1[i]
                keys.append(swap2[i])
            
        c+=1
from pprint import pprint
#pprint(dic)
print(main_code)

print(code)
keys = sorted(keys, key = len )
keys = reversed(keys)
#print(keys)
for key in keys:
    code = code.replace(key, dic[key])
#print(code.replace(';', ';\n').replace('}', '}\n'))

In [ ]:
import json
from pprint import pprint
def read(path):
    return open(path, 'r', encoding = 'utf-8').read()

def reverse_code(path):
    code = read(path)
    mp = read(path+'.map')
    c = 0
    keys = []
    swap1 = []
    swap2 = []
    dic = {}
    for val in mp.split('\n'):
        if len(val)>0:
            if c%2==0:
                swap1 = val.split(',')
            else:
                swap2 = val.split(',')
                if len(swap1) != len(swap2):
                    print("map length mismatch ")
                for i in range(len(swap2)):
                    #code = code.replace(swap2[i], swap1[i])
                    dic[swap2[i]] = swap1[i]
                    keys.append(swap2[i])

            c+=1
    vals = keys.copy()
    #print(vals)
    keys = list(sorted(keys, key = len ))
    keys = list(reversed(keys))
    #print(keys)
    for key in keys:
        code = code.replace(key, dic[key])
    return code, dic, keys
    
    
def getJsonData(JsonFile):
    with open(JsonFile, encoding="utf8") as f:
        data = json.load(f)
    return data
unique_data = getJsonData('unique_data.json')
print(unique_data[0], len(unique_data))

In [ ]:
BASE="../pantho_Tufano_processing/processed/"
main_BASE = "../codes/"

data = unique_data[0]
code_file_name = data['code_file_name']
base_patch_number = data['base_patch_number']
changed_patch_number = data['changed_patch_number']
line_no = data['actual_line_number']
main_code_path = main_BASE+code_file_name+'/'+str(base_patch_number)+'.java'
print(read(main_code_path))
print("########  processed code ##############")
print(read(BASE+code_file_name+'/'+str(base_patch_number)+'.java'))
print("########  reverse code ##############")
print(reverse_code(BASE+code_file_name+'/'+str(base_patch_number)+'.java'))


In [ ]:
main_code_path = main_BASE+code_file_name+'/'+str(base_patch_number)+'.java'
processed_code_path = BASE+code_file_name+'/'+str(base_patch_number)+'.java'
#print(read(main_code_path).split('\n'))
reverse_processed_code,dic,keys = reverse_code(processed_code_path)
pprint(dic)
print(read(main_code_path))
for line in read(main_code_path).split('\n'):
    st = line
    for key in reversed(keys):
        st=st.replace(dic[key],key)
        #print(st)
    print(st)    

In [ ]:
s = read(processed_code_path)
s = s.replace(';', ';\n').replace("}", "}\n").replace("{", "{\n")
print(s ) 

In [ ]:

s = read(main_code_path).split('\n')[line_no-1]

loc = 100000
ky = ""
val = ""

print(s)

for key in keys:
    if len(key.strip())!=0 and dic[key] in s:
        ind = s.index(dic[key])
        if ind<loc:
            loc = ind
            ky = key
            val = dic[key]

print(ky)
print(val)
print(loc)

def find()

In [ ]:
def after_process(threshold = 5):
    s = read(main_code_path).split('\n')
    s = s[max(0, line_no-threshold):min(line_no+1+threshold, len(s))]
    s = "\n".join(s)
    print(s)
after_process() 

In [ ]:
pprint(dic)

In [ ]:
print(dic)

In [ ]:
import re
method_reg = r"(public|private|static|protected|abstract|native|synchronized) +([a-zA-Z0-9<>._?, ]+) +([a-zA-Z0-9_]+) *\\([a-zA-Z0-9<>\\[\\]._?, \n]*\\) *([a-zA-Z0-9_ ,\n]*) *\\{"
main_code = read(main_code_path)
print(main_code)
#print(re.compile(method_reg))
matchs = re.search(r"(public|protected|private|static|\s) +[\w\<\>\[\]]+\s+(\w+) *\([^\)]*\) *(\{?|[^;])", main_code)
print(matchs)

In [ ]:
#print(matchs[4])
print(read(main_code_path))

In [ ]:
import re
xx = "guru99,education is fun"
r1 = re.findall(r"^\w+",xx)
print(r1)

In [ ]:

main_BASE = "../codes/acumos_12"
main_code = open(main_BASE+'/2.java', 'r', encoding = 'utf-8').read()
comment_regex = r"""^([^"\r\n]*?(?:(?<=')"[^"\r\n]*?|(?<!')"[^"\r\n]*?"[^"\r\n]*?)*?)(?<!/)/\*[^\*]*(?:\*+[^/][^\*]*)*?\*+/"""
import re
xx = main_code
print(xx)
for i in range(10):
    #xx = re.sub(r"""^([^"\r\n]*?(?:(?<=')"[^"\r\n]*?|(?<!')"[^"\r\n]*?"[^"\r\n]*?)*?)(?<!/)/\*[^\*]*(?:\*+[^/][^\*]*)*?\*+/""", "",xx)
    xx = re.sub(r"""\/\*([\S\s]+?)\*\/""", "", xx)
print(xx)

In [ ]:
$1:^([^"\r\n]*?(?:(?<=')"[^"\r\n]*?|(?<!')"[^"\r\n]*?"[^"\r\n]*?)*?)(?<!/)/\*[^\*]*(?:\*+[^/][^\*]*)*?\*+/

In [ ]:
from os import path
import json
def getJsonData(JsonFile):
    with open(JsonFile, encoding="utf8") as f:
        data = json.load(f)
    return data

def read(path):
    return open(path, 'r', encoding = 'utf-8').read()

unique_data = getJsonData('unique_data.json')
data_count = 0 
BASE="../pantho_Tufano_processing/method_data_processed/"
for data in unique_data:
    code_file_name = data['code_file_name']
    base_patch_number = data['base_patch_number']
    main_code_path = BASE+code_file_name+'/'+str(base_patch_number)+'.java'
    #print(main_code_path)
    if path.exists(main_code_path):
        data_count+=1
print(data_count)

In [ ]:
#code _summary
BASE="../pantho_Tufano_processing/codes/"
data = unique_data[1020]
code_file_name = data['code_file_name']
base_patch_number = data['base_patch_number']
main_code_path = BASE+code_file_name+'/'+str(base_patch_number)+'.java'
main_code = read(main_code_path)
print(main_code)
matchs = re.findall(r"(public|protected|private|static|\s) +([\w\<\>\[\]]+\s+(\w+)) *\([^\)]*\) *(\{?|[^;])", main_code)
print(matchs)

In [ ]:
#print(main_code)
matchs = re.findall(r"(?:(public\s))?(class|interface|enum)\s([^\n\s]*)", main_code)
print(matchs)

In [ ]:
matches = re.findall(r""""[^"]*"|((?=_[a-z_0-9]|[a-z])[a-z_0-9]+(?=\s*=))""", read('demo.java'))
print(matches)
print(re.findall(r"""([a-zA-Z][a-zA-Z0-9_$]+)\s*\(\s*([a-zA-Z0-9_$]+(?:\s*,\s*[a-zA-Z0-9_$]+)*)\s*\)""", read('demo.java')))

In [365]:
#print(read("demo.java"))
get_source_target("E:/pantho_Tufano_processing/codes/acumos_24/1.java", "E:/pantho_Tufano_processing/codes/acumos_24/2.java", 10, 1)

('E:\\pantho_Tufano_processing\r\n\n * ===============LICENSE_START=======================================================\n * Acumos\n * ===================================================================================\n * Copyright (C) 2017 AT&T Intellectual Property & Tech Mahindra. All rights reserved.\n * ===================================================================================\n * This Acumos software file is distributed by AT&T and Tech Mahindra\n * under the Apache License, Version 2.0 (the "License");\n * you may not use this file except in compliance with the License.\n<|startfocus|>\n * You may obtain a copy of the License at\n<|endfocus|>\n *  \n *      http://www.apache.org/licenses/LICENSE-2.0\n *  \n * This file is distributed on an "AS IS" BASIS,\n * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n * See the License for the specific language governing permissions and\n * limitations under the License.\n * ===============LICENSE_END=

In [ ]:
matchs = re.findall(r"(public|protected|private|static|\s) +([\w\<\>\[\]]+\s+(\w+)) *\([^\)]*\) *(\{?|[^;])", read('demo.java'))
print(matchs)

In [ ]:
sample = read('demo.java')
#print(sample)

BASE="../codes/"
data = unique_data[10]
code_file_name = data['code_file_name']
base_patch_number = data['base_patch_number']
main_code_path = BASE+code_file_name+'/'+str(base_patch_number)+'.java'
main_code = read(main_code_path)
print(main_code)

sample = main_code
for i in range(10):
    sample = re.sub(r"""\/\*([\S\s]+?)\*\/""", "", sample)
#print(sample)
classes = re.findall(r"(?:(public\s))?(class)\s([^\n\s]*)", sample)
#print(classes)
#print(classes)

def end_scope(start_index, lines):
    #print(lines)
    counter = 0
    end_index = -1
    for index in range(start_index, len(lines), 1):
        #print(lines[index])
        #print(counter)
        for char in lines[index]:
            if(char=='{'):
                counter+=1
            elif(char == "}"):
                counter-=1
            if(counter==0 and index!=start_index):
                end_index = index
                break
        if end_index!=-1:
            break
    return end_index
    
lines = sample.split('\n')
class_list = []
line_no  = 5
for cls_tpl in classes:
    s = "".join(cls_tpl[-1])
    s = s.strip()
    class_list.append(s) ## class list
    cls = s
    print(s)
    start_index = -1
    for i in range(len(lines)):
        if cls in lines[i]:
            start_index = i
            break
    print(start_index)
    end_index = end_scope(start_index, lines)
    print(end_index)
    class_scope_dic = {}
    for i in range(start_index, end_index+1):
        class_scope_dic[i] = True
    funcs = re.findall(r"(public|protected|private|static|\s) +([\w\<\>\[\]]+\s+(\w+)) *\([^\)]*\) *(\{?|[^;])", "\n".join(lines[start_index:end_index+1]))
    func_list = []
    func_scopes = []
    #print(funcs)
    special_func = ""
    for func in funcs:
        #print(func)
        if(func[-1]!= "{"):
            continue
        fc = list(func)
        fc = fc[-3]
        s = "".join(fc)
        s = s.strip()
        #func_list.append(s) ## function list
        #print(func_list)
        for index in range(start_index, end_index+1, 1):
            if s in lines[index]:
                #print("start func = ", index)
                #print("end func = ", end_scope(index, lines))
                #print("#####", lines[index].strip())
                start_func = index
                end_func = end_scope(index, lines)
                func_list.append(lines[index].strip())
                func_scopes.append((start_func,end_func))
                if(start_func<=line_no<=end_func):
                    special_func = "\n".join(lines[start_func:end_func+1])
                    print(special_func)
                break
    print(func_list)
    for func_sc in func_scopes:
        for i in range(func_sc[0], func_sc[1]+1):
            class_scope_dic[i] = False
    
    prime_var_list = []
    dic_vars = {}
    for i in range(start_index, end_index+1):
        if class_scope_dic[i]==True:
            #prime_vars = re.findall(r""""[^"]*"|((?=_[a-z_0-9]|[a-z])[a-z_0-9]+((?=\s*=)))""",lines[i])
            if '(' not in lines[i] and "return" not in lines[i] and "extends" not in lines[i]:
                prime_vars = re.findall(r"""(\w+\s+)([a-zA-Z_][a-zA-Z0-9_]*)""", lines[i])
                #print(lines[i])
                #print(prime_vars)
                #print(prime_vars[1][1])
                if len(prime_vars)==2:
                    #print(prime_vars)
                    if(len(prime_vars[1])==2 and prime_vars[1][1] in dic_vars.keys()):
                        dic_vars[prime_vars[1][1]]+=1
                    elif len(prime_vars[1])==2:
                        dic_vars[prime_vars[1][1]] =1
    if(len(list(dic_vars.keys()))>0):
        print(dic_vars)
    #print(lines[i])
    '''
            for var in prime_vars:
                for sample in var:
                    if len(sample.replace(';', ''))>0:
                        got_var = sample.replace(';', '')
                        if(got_var in lines[i].split(" ")):
                            print("var = ", got_var)
    '''

In [ ]:
s = "private final Map<LeScanCallback, ScanCallback> mLeScanClients;"
print(s.split(" ")[-1])

In [344]:
from nltk.tokenize import MWETokenizer, TweetTokenizer, WordPunctTokenizer 
from nltk.tokenize import word_tokenize
import re

VALUES = ['>>>=', '>>=', '<<=',  '%=', '^=', '|=', '&=', '/=',
                  '*=', '-=', '+=', '<<', '--', '++', '||', '&&', '!=',
                  '>=', '<=', '==', '%', '^', '|', '&', '/', '*', '-',
                  '+', ':', '?', '~', '!', '<', '>', '=', '...', '->', '::', '\\', '\\\\', '\*', '*\\', '\\\\\\']
INFIX = ['||', '&&', '|', '^', '&', '==', '!=', '<', '>', '<=', '>=', 
             '<<', '>>', '>>>', '+', '-', '*', '/', '%']
PREFIX = ['++', '--', '!', '~', '+', '-']
POSTFIX = ['++', '--']
ASSIGNMENT = ['=', '+=', '-=', '*=', '/=', '&=', '|=', '^=', '%=', '<<=', '>>=', '>>>=']
LAMBDA = ['->']
COMMENT = ['//', '/*', '*/']
METHOD_REFERENCE = ['::']

WHITESPACE_DICT = {"                ":'<|16-s|>', 
                   "            ":'<|12-s|>', 
                   "        ":'<|8-s|>', 
                   "    ":'<|4-s|>', 
                   "  ":'<|2-s|>', " ":'<|s|>',
                   "\t\t\t\t":'<|4-t|>',"\t\t\t":'<|3-t|>',"\t\t":'<|2-t|>',"\t":'<|t|>',"\n":'<|nl|>'}

WHITESPACES = [WHITESPACE_DICT[x] for x in WHITESPACE_DICT]

REVERSE_WHITESPACE_DICT = {}
for key in WHITESPACE_DICT:
    value = WHITESPACE_DICT[key]
    REVERSE_WHITESPACE_DICT[value] = key

CUSTOM_TOKEN = ["<|startcode|>", "<|endcode|>", "<|startfocus|>", "<|endfocus|>", "<|startcomment|>", "<|endcomment|>", 
                   "<|stringliteral|>", "<|singlelinecomment|>", "<|multilinecomment|>", "<|del|>"]

values = list(set(INFIX+PREFIX+POSTFIX+ASSIGNMENT+LAMBDA+COMMENT+METHOD_REFERENCE))
token_phrases = []
word_punct_tokenizer = WordPunctTokenizer()
for tok in CUSTOM_TOKEN+WHITESPACES:
    temp = tuple(word_punct_tokenizer.tokenize(tok))
    token_phrases.append(temp)
for w in values:
    temp = tuple(MWETokenizer().tokenize(w))
    if len(temp) > 1:
        token_phrases.append(temp)

word_punct_tokenizer = WordPunctTokenizer()
tweet_tokenizer = TweetTokenizer()
mwe_tokenizer = MWETokenizer(token_phrases, separator="")

def state(c):
    n = ord(c)
    if n>=97 and n<=122: # lower case
        return 1
    elif n>=65 and n<=90: # upper case
        return 2
    elif n>=48 and n<=57: # numbers
        return 3
    elif c.isspace(): # whitespaces
        return 4
    elif c in ['_', '$']: 
        return 5
    elif n < 128:
        return 6
    else:
        return 7

def space_up(s):
    if s is None or s == "":
        return ""
    new_s = s[0]
    for i in range(1,len(s)):
        prev_state = state(s[i-1])
        curr_state = state(s[i])
        if prev_state in [1,2] and curr_state in [3]:
            new_s += " "
        elif prev_state in [1] and curr_state in [2]:
            new_s += " "
        elif prev_state in [3] and curr_state in [1,2]:
            new_s += " "
        elif prev_state in [1,2,3] and curr_state in [5]:
            new_s += " "
        elif prev_state in [5] and curr_state in [1,2,3]:
            new_s += " "
        new_s+=s[i]
    return new_s

def white_space_tokenize(s):
    for x in WHITESPACE_DICT:
        s = s.replace(x, WHITESPACE_DICT[x])
    for key in REVERSE_WHITESPACE_DICT:
        #val = REVERSE_WHITESPACE_DICT[x]
        s = s.replace(key, " "+key+" ")
    return s

def extreme_tokenization(comment):
    comment = white_space_tokenize(comment)
    comment = space_up(comment)
    tokenized = tweet_tokenizer.tokenize(comment)
    tokenized = word_punct_tokenizer.tokenize(' '.join(tokenized))
    tokenized = mwe_tokenizer.tokenize(tokenized)
    tokenized_comment = ' '.join(tokenized)
    tokenized_comment = re.sub(r'[^\x00-\x7f]',r'', tokenized_comment)
    tokenized = tokenized_comment.split()
    return tokenized

In [357]:
code = """
if (cornoVirus< 5){  } """
#code = white_space_tokenize(code)
#code = space_up(code)
#code = tweet_tokenizer.tokenize(code)
print(code)


if (cornoVirus< 5){  } 


In [359]:
code = extreme_tokenization(code)
print(code)

['<|nl|>', 'if', '<|s|>', '(', 'corno', 'Virus', '<', '<|s|>', '5', ')', '{', '<|2-s|>', '}', '<|s|>']


In [360]:
print(extreme_detokenization(code))


if (cornoVirus< 5){  } 


In [363]:
from os import path
import json
import re
import subprocess
import nltk
from pprint import pprint
from nltk.tokenize import sent_tokenize, word_tokenize
from tqdm import tqdm
nltk.download('punkt')

def getJsonData(JsonFile):
    with open(JsonFile, encoding="utf8") as f:
        data = json.load(f)
    return data
def read(path):
    return open(path, 'r', encoding = 'utf-8').read()


BASE="E:/codes/"
unique_data = getJsonData('E:\\unified.json')
data = unique_data[1020]
code_file_name = data['file_name']
base_patch_number = data['base_patch_number']
main_code_path = BASE+code_file_name+'/'+str(base_patch_number)+'.java'
main_code = read(main_code_path)

# ERROR: EVERY OUTPUT IS ACCOMPANIED BY THE PATH IN THE CMD OUTPUT. WE NEEDED TO REMOVE THAT MANUALLY.

def get_source_target(file1, file2, line_number, change_window_size=5):
    source_target = subprocess.check_output(['java', '-jar', 'ChangedLineF.jar', file1, file2, str(line_number), str(change_window_size)]).decode("utf-8")
    (source, target) = source_target.split("<|source-target-sep|>")
    return source, target

def end_scope(start_index, lines):
    #print(lines)
    counter = 0
    end_index = -1
    found = False
    for index in range(start_index, len(lines), 1):
        #print(lines[index])
        #print(counter)
        for char in lines[index]:
            if(char=='{'):
                counter+=1
                found = True
            elif(char == "}"):
                counter-=1
                found = True
            if(counter==0 and found):
                end_index = index
                break
        if end_index!=-1:
            break
    return end_index

unique_data_processed = []
MAX_NUM_TOKEN = 800
UP_TOKEN = MAX_NUM_TOKEN//2
c = 0
ec = 0
errors = []
for idx, data in enumerate(unique_data):#[2428:2429]#[1000:1001]
    c+=1
    code_file_name = data['file_name']
    base_patch_number = data['base_patch_number']
    changed_patch_number = data['changed_patch_number']
    main_code_path = BASE+code_file_name+'/'+str(base_patch_number)+'.java'
    changed_code_path = BASE+code_file_name+'/'+str(changed_patch_number)+'.java'
    line_no = data['line_number']
    main_code = read(main_code_path)
    changed_code = read(changed_code_path)
    sample = main_code
    message = data['message']
    comment_id = data['comment_id']
    print(read(main_code_path))
    try:
        source, target = get_source_target(main_code_path, changed_code_path,line_no-1)
    except:
        error = {}
        error['main_code'] = main_code
        error['changed_code'] = changed_code
        error['line_no'] = line_no
        ec+=1
        if(main_code!=''):
            errors.append(error)
        print("falied to store unique_data= ", c, "error count", ec)    
    print(source)
    break
    '''
    if len(errors)>100:
        with open('errors.json', 'w', encoding="utf8") as f:  # writing JSON object
            json.dump(errors, f)
        break
    '''
    #print(source)
    #print(len(errors))
    #print(ec)
    #continue
    sample = source
    #print(source)
    #for i in range(10):
    #    sample = re.sub(r"""\/\*([\S\s]+?)\*\/""", "", sample)
    
    data_dic = {}
    
    #data_dic['source'] = source
    data_dic['message'] = message
    
    data_dic['comment_id'] = comment_id
    data_dic['target'] = target
    data_dic['source'] = source
    data_dic['idx'] = c
    func_list = []
    data_dic['code_snippet'] = ""
    data_dic['prime_var_dic'] = {}
    #print(sample)
    classes = re.findall(r"(?:(public\s))?(class)\s([^\n\s]*)", sample)
    lines = sample.split('\n')
    class_list = []
    sp_start_func= -1
    sp_end_func= len(sample.split('\n'))   
    for cls_tpl in classes:
        s = "".join(cls_tpl[-1])
        s = s.strip()
        class_list.append(s) ## class list
        cls = s
        #print(s)
        start_index = -1
        for i in range(len(lines)):
            if cls in lines[i]:
                start_index = i
                break
        #print(start_index)
        end_index = end_scope(start_index, lines)
        #print(end_index)
        class_scope_dic = {}
        for i in range(start_index, end_index+1):
            class_scope_dic[i] = True
        funcs = re.findall(r"(public|protected|private|static|\s) +([\w\<\>\[\]]+\s+(\w+)) *\([^\)]*\) *(\{?|[^;])", "\n".join(lines[start_index:end_index+1]))
        func_scopes = []
        #print(funcs)
        #continue
        visited_func = set()
        for func in funcs:
            #print(func)
            if(func[-1]!= "{"):
                continue
            fc = list(func)
            fc = fc[-3]
            s = "".join(fc)
            s = s.strip()
            #func_list.append(s) ## function list
            #print(func_list)
            
            for index in range(start_index, end_index+1, 1):
                if s in lines[index]:
                    start_func = index
                    end_func = end_scope(index, lines)
                    
                    func_scopes.append((start_func,end_func))
                    if s not in visited_func:
                        visited_func.add(s)
                        func_list.append(lines[index].replace("{", "").strip())
                    
                    if(start_func<=line_no<=end_func):
                        special_func = "\n".join(lines[start_func:end_func+1])
                        #data_dic['code_snippet'] = special_func
                        #nonlocal sp_start_func
                        sp_start_func= start_func
                        #nonlocal sp_end_func
                        sp_end_func= end_func
        
        for func_sc in func_scopes:
            for i in range(func_sc[0], func_sc[1]+1):
                class_scope_dic[i] = False
        prime_var_list = []
        dic_vars = {}
        for i in range(start_index, end_index+1):
            if class_scope_dic[i]==True:
                #prime_vars = re.findall(r""""[^"]*"|((?=_[a-z_0-9]|[a-z])[a-z_0-9]+((?=\s*=)))""",lines[i])
                if '(' not in lines[i] and "return" not in lines[i] and "extends" not in lines[i]:
                    prime_vars = re.findall(r"""(\w+\s+)([a-zA-Z_][a-zA-Z0-9_]*)""", lines[i])
                    if len(prime_vars)==2:
                        if(len(prime_vars[1])==2 and prime_vars[1][1] in dic_vars.keys()):
                            dic_vars[prime_vars[1][1]]+=1
                        elif len(prime_vars[1])==2:
                            dic_vars[prime_vars[1][1]] =1
        #if(len(list(dic_vars.keys()))>0):
            #print(dic_vars)
        data_dic['prime_var_dic'] = dic_vars
    data_dic['class_list'] = class_list
    data_dic['func_list'] = func_list
    
    #print("specials   ###", sp_start_func, sp_end_func)
    
    if not (0<len(extreme_tokenization(data_dic['code_snippet']))<MAX_NUM_TOKEN): 
        up_count = 0
        up_done = False
        down_count = 0
        down_done = False
        splitted = sample.split("\n")
        while(1):
            #print(max(0, line_no-up_count),line_no+1)
            if len(extreme_tokenization("\n".join(splitted[max(0,sp_start_func, line_no-up_count):line_no+1])))<UP_TOKEN:
                if(up_count==line_no):
                    break
                up_count+=1
            else:
                break
                #print("something is wrong")
                      
        while(1):
            if len(extreme_tokenization("\n".join(splitted[(line_no-up_count):min(line_no+down_count+1,sp_end_func, len(splitted))])))<MAX_NUM_TOKEN:
                if(down_count==len(splitted)):
                    break
                down_count+=1
            else:
                break
    if sp_start_func != -1:
        #data_dic['code_snippet'] = "\n".join(splitted[sp_start_func:sp_end_func+1])
        
        if("startfocus" in splitted[sp_start_func-1]):
            sp_start_func-=1
        if sp_end_func <=len(splitted)-2 and "endfocus" in splitted[sp_end_func+1]:
            sp_end_func+=2
        if sp_end_func <=len(splitted)-1 and "endfocus" in splitted[sp_end_func]:
            sp_end_func+=1
        #print(" ".join(splitted[sp_start_func:sp_end_func+1]))
        #print(sp_start_func)
        #print(sp_end_func)
        code_snippet = "\n".join(splitted[sp_start_func:sp_end_func+1])
        print("in function")
        
    else:
        #data_dic['code_snippet'] = "\n".join(splitted[(line_no-up_count):(line_no+down_count+1)])
        print("not in function")
        if("startfocus" in splitted[(line_no-up_count-1)]):
            up_count+=1
        if line_no+down_count <=len(splitted)-1 and "endfocus" in splitted[line_no+down_count]:
            down_count+=1
        code_snippet = "\n".join(splitted[(line_no-up_count):(line_no+down_count+1)])
        #print(line_no-up_count)
        #print(line_no+down_count+1)
        #print(code_snippet)
    #print(code_snippet)
    #print(data_dic['func_list'])
    if "<|startfocus|>" in code_snippet and "<|endfocus|>" not in code_snippet:
        code_snippet =code_snippet+  "\n<|endfocus|>"
    if "<|startfocus|>" not in code_snippet and "<|endfocus|>" in code_snippet:
        code_snippet = "<|startfocus|>\n"+code_snippet
        
    data_dic['tokenized_code_snippet'] = extreme_tokenization(code_snippet)#data_dic['code_snippet'])
    data_dic['tokenized_target'] = extreme_tokenization(data_dic['target'])
    data_dic['tokenized_comment'] = extreme_tokenization(data_dic['message'])
    #print("specials   ###", sp_start_func, sp_end_func)
    if(idx%1000==0):
        print(idx)
        print(c)
        pprint(data_dic)
    
    #print(code_snippet)
    #print("#################")
    #print(data_dic)
    #print(source)
    #data_dic['actual_line_number'] = data['line_number']
    data_dic['base_code_line_number'] = data['line_number']
    data_dic['base_patch_number'] = data['base_patch_number']
    data_dic['changed_patch_number'] = data['changed_patch_number']
    data_dic['code_file_name'] = data['file_name']
    data_dic['comment_id'] = data['comment_id']
    data_dic['message'] = data['message']
    data_dic['line_change'] = data['line_change']
    if "<|startfocus|>" in code_snippet and "<|endfocus|>" in code_snippet:
        unique_data_processed.append(data_dic)
    print(idx)
    
'''
{'actual_line_number': 71, 'base_code_line_number': 73, 'base_patch_number': 3, 'changed_code': '    /*', 'changed_patch_number': 7, 'code_file_name': 'android_3478', 'comment_id': '3745e284_1e49cdaa', 'line_change': 2, 'message': '{@link android.icu.impl.OlsonTimeZone} ?', 'previous_code': '     /*'}
'''

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\HP\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!



 * ===============LICENSE_START=======================================================
 * Acumos
 * ===================================================================================
 * Copyright (C) 2017 AT&T Intellectual Property & Tech Mahindra. All rights reserved.
 * ===================================================================================
 * This Acumos software file is distributed by AT&T and Tech Mahindra
 * under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *  
 *      http://www.apache.org/licenses/LICENSE-2.0
 *  
 * This file is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ===============LICENSE_END=========================================================
 */

package org.acumos.

"\n{'actual_line_number': 71, 'base_code_line_number': 73, 'base_patch_number': 3, 'changed_code': '    /*', 'changed_patch_number': 7, 'code_file_name': 'android_3478', 'comment_id': '3745e284_1e49cdaa', 'line_change': 2, 'message': '{@link android.icu.impl.OlsonTimeZone} ?', 'previous_code': '     /*'}\n"

In [315]:
with open('unique_data_processed_v2_withoutDate.json', 'w', encoding="utf8") as f:  # writing JSON object
    json.dump(unique_data_processed, f)

In [ ]:
def getJsonData(JsonFile):
    with open(JsonFile, encoding="utf8") as f:
        data = json.load(f)
    return data

In [319]:
dt = getJsonData('unique_data_processed_v2_withoutDate.json')
print(dt[0])

{'message': 'One last occurrence - please change to acumos.org/...', 'comment_id': 'a068aa5b_b404cd60', 'target': '\t\t\t\t\t\t"https://cognita.atlassian.net/wiki/spaces/CW/pages/233610/Data+Model+for+Machine+Learning+Platform",\n\t\t\t\t\t\t"cds.someday@acumos.org"), // Contact\n\r\n', 'source': 'E:\\pantho_Tufano_processing\r\n\n * ===============LICENSE_START=======================================================\n * Acumos\n * ===================================================================================\n * Copyright (C) 2017 AT&T Intellectual Property & Tech Mahindra. All rights reserved.\n * ===================================================================================\n * This Acumos software file is distributed by AT&T and Tech Mahindra\n * under the Apache License, Version 2.0 (the "License");\n * you may not use this file except in compliance with the License.\n * You may obtain a copy of the License at\n *  \n *      http://www.apache.org/licenses/LICENSE-2.0\n * 

In [108]:
unique_data_processed[2020]

{'message': 'no space',
 'comment_id': '6775186c_dfa65451',
 'target': '        if (!(methodRef instanceof CstMethodRef)) {\n\r\n',
 'idx': 2021,
 'code_snippet': '    public boolean isCompatible(DalvInsn insn) {\n        if (!(insn instanceof MultiCstInsn)) {\n            return false;\n        }\n\n        MultiCstInsn mci = (MultiCstInsn) insn;\n        if (mci.getNumberOfConstants() != 2) {\n            return false;\n        }\n\n        int methodIdx = mci.getIndex(0);\n        int protoIdx = mci.getIndex(1);\n        if (! unsignedFitsInShort(methodIdx) || ! unsignedFitsInShort(protoIdx)) {\n            return false;\n        }\n\n        Constant methodRef = mci.getConstant(0);\n<|startfocus|>\n        if (! (methodRef instanceof CstMethodRef)) {\n<|endfocus|>\n            return false;\n        }\n\n        Constant protoRef = mci.getConstant(1);\n        if (! (protoRef instanceof CstProtoRef)) {\n            return false;\n        }\n\n        RegisterSpecList regs = mci.get

In [113]:
for i in range(len(unique_data_processed)):
    #unique_data_processed[i]['tokenized_comment'] = extreme_tokenization(unique_data_processed[i]['message']) 
    del unique_data_processed[i]['code_snippet']

In [117]:
for x in unique_data_processed[100]:
    print(x)

message
comment_id
target
idx
prime_var_dic
class_list
func_list
tokenized_code_snippet
tokenized_target
tokenized_comment


## Remove Duplicates

In [176]:
def merge_code_comment_old(x):
    message = x['message']
    changed_code = x['changed_code']
    previous_code = x['previous_code']
    merged = message + '<>' + changed_code + '<>' + previous_code
    return merged

In [316]:
def merge_code_comment(x):
    message = x['message']
    source = ' '.join(x['tokenized_code_snippet'])
    target = ' '.join(x['target'])
    merged = message + '<>' + source + '<>' + target
    return merged

In [317]:
import traceback
existing_ids = set()
duplicate_ids = set()
unique_data = []
i = 0
target_size = 400
error_count = 0
error_list = []

for x in unique_data_processed:
    i+= 1
    try:
        focuslen = x['tokenized_code_snippet'].index('<|endfocus|>') - x['tokenized_code_snippet'].index('<|startfocus|>')
        key = merge_code_comment(x)
        if key not in existing_ids and len(x['tokenized_target']) <= target_size and focuslen <= target_size:
            existing_ids.add(key)
            unique_data.append(x)
        elif key not in duplicate_ids:
            duplicate_ids.add(key)
    except Exception as e:
        print(e)
        x['error'] = e
        error_list.append(x)
        error_count += 1
        #print(x)
        #break
print(len(unique_data_processed))
print(len(unique_data))

88316
78321


In [298]:
pprint(error_list[0])
#bcee4497_c46a9bac

{'actual_line_number': 370,
 'base_code_line_number': 370,
 'base_patch_number': 1,
 'changed_patch_number': 2,
 'class_list': ['performs', 'AccessibilityEndToEndTest'],
 'code_file_name': 'android_5244',
 'code_snippet': '',
 'comment_id': 'bcee4497_c46a9bac',
 'error': ValueError("'<|endfocus|>' is not in list"),
 'func_list': ['public AccessibilityEndToEndTest()',
               'new Runnable()',
               'public void run()',
               'public boolean accept(AccessibilityEvent event)',
               'private boolean equalsAccessiblityEvent(AccessibilityEvent '
               'first, AccessibilityEvent second)',
               'private boolean '
               'equalsNotificationAsParcelableData(AccessibilityEvent first,',
               'private boolean equalsText(AccessibilityEvent first, '
               'AccessibilityEvent second)'],
 'idx': 171,
 'line_change': 2,
 'message': "I'm not intimately familiar with notificationManager. Is there a "
            'potential r

           '                // trigger the event\n'
           '                getActivity().runOnUiThread(new Runnable() {\n'
           '                    @Override\n'
           '                    public void run() {\n'
           '                        button.performClick();\n'
           '                    }\n'
           '                });\n'
           '            }},\n'
           '            new UiAutomation.AccessibilityEventFilter() {\n'
           '                // check the received event\n'
           '                @Override\n'
           '                public boolean accept(AccessibilityEvent event) {\n'
           '                    return equalsAccessiblityEvent(event, '
           'expected);\n'
           '                }\n'
           '            },\n'
           '            TIMEOUT_ASYNC_PROCESSING);\n'
           '        assertNotNull("Did not receive expected event: " + '
           'expected, awaitedEvent);\n'
           '    }\n'
    

           '        AccessibilityEvent awaitedTextChangeEvent =\n'
           '            '
           'getInstrumentation().getUiAutomation().executeAndWaitForEvent(\n'
           '                new Runnable() {\n'
           '            @Override\n'
           '            public void run() {\n'
           '                // trigger the event\n'
           '                getActivity().runOnUiThread(new Runnable() {\n'
           '                    @Override\n'
           '                    public void run() {\n'
           '                        editText.getEditableText().replace(3, 4, '
           'newText);\n'
           '                    }\n'
           '                });\n'
           '            }},\n'
           '            new UiAutomation.AccessibilityEventFilter() {\n'
           '                // check the received event\n'
           '                @Override\n'
           '                public boolean accept(AccessibilityEvent event) {\n'
        

           '                @Override\n'
           '                public boolean accept(AccessibilityEvent event) {\n'
           '                    return equalsAccessiblityEvent(event, '
           'expected);\n'
           '                }\n'
           '            },\n'
           '            TIMEOUT_ASYNC_PROCESSING);\n'
           '        assertNotNull("Did not receive expected event: " + '
           'expected, awaitedEvent);\n'
           '    }\n'
           '\n'
           '    /**\n'
           '     * Compares all properties of the <code>first</code> and the\n'
           '     * <code>second</code>.\n'
           '     */\n'
           '    private boolean equalsAccessiblityEvent(AccessibilityEvent '
           'first, AccessibilityEvent second) {\n'
           '         return first.getEventType() == second.getEventType()\n'
           '            && first.isChecked() == second.isChecked()\n'
           '            && first.getCurrentItemIndex() == '
         

                            'finish',
                            '(',
                            ');',
                            '<|nl|>',
                            '<|16-s|>',
                            '<|4-s|>',
                            '}'],
 'tokenized_comment': ['I',
                       "'",
                       'm',
                       '<|s|>',
                       'not',
                       '<|s|>',
                       'intimately',
                       '<|s|>',
                       'familiar',
                       '<|s|>',
                       'with',
                       '<|s|>',
                       'notification',
                       'Manager',
                       '.',
                       '<|s|>',
                       'Is',
                       '<|s|>',
                       'there',
                       '<|s|>',
                       'a',
                       '<|s|>',
                       'potential',
             

In [215]:
error_list[2]['tokenized_code_snippet'].index('<|startfocus|>')

ValueError: '<|startfocus|>' is not in list

In [154]:
def extreme_detokenization(tokens):
    s = ""
    for token in tokens:
        if token in REVERSE_WHITESPACE_DICT:
            s+= REVERSE_WHITESPACE_DICT[token]
        else:
            s+= token
    for key in REVERSE_WHITESPACE_DICT:
        val = REVERSE_WHITESPACE_DICT[key]
        s = s.replace(key, value)
    return s

In [323]:
d = {"x": 1, "y": 2}
for x in d:
    print(x,d[x])

x 1
y 2


In [220]:
print(extreme_detokenization(error_list[3]['tokenized_code_snippet']))

    public boolean putClientChainInKeyStore(String name, Certificate[] certs) {
<|endfocus|>
        try {
            byte[] certData = Credentials.convertToPem(certs);
            if (mVerboseLoggingEnabled) {
                Log.d(TAG, "putting " + certs.length + " certificate(s) "
                        + name + " in keystore");
            }
            return mKeyStore.put(name, certData, Process.WIFI_UID, KeyStore.FLAG_NONE);
        } catch (IOException e1) {
            return false;
        } catch (CertificateException e2) {
            return false;
        }
    }


In [249]:
print(error_list[2])

{'message': 'Re-indentation.', 'comment_id': 'b62d1c44_17563d1c', 'target': '      mDevice.findObject(new UiSelector().text(Constants.TIP_BUTTON_OK))\n          .clickAndWaitForNewWindow();\n\r\n', 'source': 'E:\\pantho_Tufano_processing\r\n * Copyright (c) 2017 The Android Open Source Project\n *\n * Licensed under the Apache License, Version 2.0 (the "License");\n * you may not use this file except in compliance with the License.\n * You may obtain a copy of the License at\n *\n *     http://www.apache.org/licenses/LICENSE-2.0\n *\n * Unless required by applicable law or agreed to in writing, software\n * distributed under the License is distributed on an "AS IS" BASIS,\n * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n * See the License for the specific language governing permissions and\n * limitations under the License.\n */\n\npackage com.android.devtools.server.services;\n\nimport com.android.devtools.server.model.RestServiceModel;\nimport com.android

In [144]:
len(unique_data)

26815

Original Unique Data: 38268
Less than 400 token in target: 38009
Less than 100 token in target: 34975
Less than 400 token, has focus: 26815

In [124]:
len(unique_data_processed)

70571

In [125]:
len(existing_ids)

38268

In [139]:
unique_data[5]['tokenized_code_snippet'].index('<|startfocus|>')

135

In [142]:
unique_data[0]['tokenized_code_snippet'].index('<|endfocus|>')

196

In [159]:
print(len(unique_data_processed))
print(unique_data_processed[0])

70571


{'message': 'This is an ugly copyright line.  Please fix in all files.', 'comment_id': 'c80bb11f_61c30f6e', 'target': '\n * ===============LICENSE_START=======================================================\n * Acumos\n * ===================================================================================\n * Copyright (C) 2017 AT&T Intellectual Property & Tech Mahindra. All rights reserved.\n * ===================================================================================\n * This Acumos software file is distributed by AT&T and Tech Mahindra\n * under the Apache License, Version 2.0 (the "License");\n * you may not use this file except in compliance with the License.\n * You may obtain a copy of the License at\n *  \n *      http://www.apache.org/licenses/LICENSE-2.0\n *  \n * This file is distributed on an "AS IS" BASIS,\n * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n * See the License for the specific language governing permissions and\n * limitat

In [250]:

for idx, data in enumerate(unique_data_processed):
    if data['comment_id'] == "b62d1c44_17563d1c":
        print(idx)
        print(data)
        break
#acd32db6_cc256763

188
{'message': 'Re-indentation.', 'comment_id': 'b62d1c44_17563d1c', 'target': '      mDevice.findObject(new UiSelector().text(Constants.TIP_BUTTON_OK))\n          .clickAndWaitForNewWindow();\n\r\n', 'source': 'E:\\pantho_Tufano_processing\r\n * Copyright (c) 2017 The Android Open Source Project\n *\n * Licensed under the Apache License, Version 2.0 (the "License");\n * you may not use this file except in compliance with the License.\n * You may obtain a copy of the License at\n *\n *     http://www.apache.org/licenses/LICENSE-2.0\n *\n * Unless required by applicable law or agreed to in writing, software\n * distributed under the License is distributed on an "AS IS" BASIS,\n * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n * See the License for the specific language governing permissions and\n * limitations under the License.\n */\n\npackage com.android.devtools.server.services;\n\nimport com.android.devtools.server.model.RestServiceModel;\nimport com.and

In [295]:
def getJsonData(JsonFile):
    with open(JsonFile, encoding="utf8") as f:
        data = json.load(f)
    return data
def read(path):
    return open(path, 'r', encoding = 'utf-8').read()
unique_data = getJsonData('unified_diff_2.json')

for idx, data in enumerate(unique_data):
    if data['comment_id'] == "bcee4497_c46a9bac":
        print(idx)
        print(data)
        break

3170
{'actual_line_number': 370, 'base_code_line_number': 370, 'base_patch_number': 1, 'changed_code': '                        return equalsAccessiblityEvent(event, expected);', 'changed_patch_number': 2, 'code_file_name': 'android_5244', 'comment_id': 'bcee4497_c46a9bac', 'line_change': 2, 'message': "I'm not intimately familiar with notificationManager. Is there a potential race condition here, where the notification could be canceled before it is reported to accessibility? If there isn't an ironclad guarantee that the notification will always get through even if it's been canceled, I would move the cancelation into a finally block so it's done at the end of the test.", 'previous_code': '                        getActivity().finish();'}


In [309]:
def getJsonData(JsonFile):
    with open(JsonFile, encoding="utf8") as f:
        data = json.load(f)
    return data
def read(path):
    return open(path, 'r', encoding = 'utf-8').read()
unique_data = getJsonData('unified_diff_2.json')
dat = getJsonData('E:\\unified.json')

In [310]:
print(dat[0])

{'base_patch_number': 3, 'changed_patch_number': 4, 'comment_id': 'a068aa5b_b404cd60', 'file_name': 'acumos_197', 'line_change': 0, 'line_number': 64, 'message': 'One last occurrence - please change to acumos.org/...'}
